pkg/provider/provider.go

```diff
@@ -1231,3 +1266,178 @@ func parseLiveInputs(live, oldInputs *unstructured.Unstructured) *unstructured.U
+// - If the types of the old and new values differ, the property is recorded as updated
+// - If both values are maps, the maps are recursively compared on a per-property basis and added to the diff
+// - If both values are arrays, the arrays are recursively compared on a per-element basis and added to the diff
+// - If both values are primitives and the values differ, the property is recorded as updated
+// - Otherwise, no diff is recorded.
+//
+// If a difference is present at the given path and the path matches one of the patterns in the database of
+// force-new properties, the diff is amended to indicate that the resource needs to be replaced due to the change in
+// this property.
+func (pc *patchConverter) addPatchValueToDiff(path []interface{}, v, old interface{}, inArray bool) error {
+	contract.Assert(v != nil || old != nil)
+
+	var diffKind pulumirpc.PropertyDiff_Kind
+	inputDiff := false
+	if v == nil {
+		diffKind, inputDiff = pulumirpc.PropertyDiff_DELETE, true
+	} else if old == nil {
+		diffKind = pulumirpc.PropertyDiff_ADD
+	} else {
+		switch v := v.(type) {
+		case map[string]interface{}:
+			if oldMap, ok := old.(map[string]interface{}); ok {
+				return pc.addPatchMapToDiff(path, v, oldMap, inArray)
+			}
+			diffKind = pulumirpc.PropertyDiff_UPDATE
+		case []interface{}:
+			if oldArray, ok := old.([]interface{}); ok {
+				return pc.addPatchArrayToDiff(path, v, oldArray, inArray)
+			}
+			diffKind = pulumirpc.PropertyDiff_UPDATE
+		default:
+			if reflect.DeepEqual(v, old) {
+				// From RFC 7386 (the JSON Merge Patch spec):
+				//
+				//   If the patch is anything other than an object, the result will always be to replace the entire
+				//   target with the entire patch. Also, it is not possible to patch part of a target that is not an
+				//   object, such as to replace just some of the values in an array.
+				//
+				// Because JSON merge patch does not allow array elements to be updated--instead, the array must be
+				// replaced in full--the patch we have is an overestimate of the properties that changed. As such, we
+				// only record updates for values that have in fact changed.
+				return nil
+			}
+			diffKind = pulumirpc.PropertyDiff_UPDATE
+		}
+	}
+
+	// Determine if this change causes a replace.
+	matches, err := openapi.PatchPropertiesChanged(makePatchSlice(path, v).(map[string]interface{}), pc.forceNew)
+	if err != nil {
+		return err
+	}
+	if len(matches) != 0 {
+		switch diffKind {
+		case pulumirpc.PropertyDiff_ADD:
+			diffKind = pulumirpc.PropertyDiff_ADD_REPLACE
+		case pulumirpc.PropertyDiff_DELETE:
+			diffKind = pulumirpc.PropertyDiff_DELETE_REPLACE
+		case pulumirpc.PropertyDiff_UPDATE:
+			diffKind = pulumirpc.PropertyDiff_UPDATE_REPLACE
+		}
+	}
+
+	pathStr := ""
+	for _, v := range path {
+		switch v := v.(type) {
+		case string:
+			if strings.ContainsAny(v, `."[]`) {
+				pathStr = fmt.Sprintf(`%s["%s"]`, pathStr, strings.ReplaceAll(v, `"`, `\"`))
```

Comment:
```suggestion
				pathStr = fmt.Sprintf(`%s[%q]`, pathStr, strings.ReplaceAll(v, `"`, `\"`))
```

Reply:
`%q` may add additional escape sequences that are not supported by the path syntax we use in the detailed diff (documented here: https://github.com/pulumi/pulumi/blob/master/sdk/proto/provider.proto#L137-L168).